src/CommunicationPattern/CommunicationPattern.H

```diff
@@ -0,0 +1,189 @@
+ * @param padding_after Number of unused elements after
+ * @return The CommCounts
+ * */
+CommCounts compute_gather_to_owner_counts(const ExecutorHandler &exec_handler,
+                                          label ranks_per_owner, label size,
+                                          label total_size,
+                                          label padding_before,
+                                          label padding_after);
+
+/* @brief  This function computes the send and recv counts and offset vectors
+ *for repartitioning to an owner rank
+ *
+ * @param exec_handler The executor handler
+ * @param ranks_per_owner Ratio of total ranks to owner ranks
+ * @parameter size how many elements to send to owner
+ * @return The CommCounts
+ * */
+CommCounts compute_gather_to_owner_counts(const ExecutorHandler &exec_handler,
+                                          label ranks_per_owner, label size);
+
+/* @brief  This function computes the send and recv counts vectors for
+ * for communication back to neighbours from an owner rank
+ *
+ * @param exec_handler The executor handler
+ * @param ranks_per_owner Ratio of total ranks to owner ranks
+ * @parameter size how many elements to send to owner
+ * @return The CommCounts
+ * */
+CommCounts compute_scatter_from_owner_comm(const ExecutorHandler &exec_handler,
+                                           label ranks_per_owner, label size);
+
+/** @brief Given a rank id and ranks_per_owners this function computes the
+ * corresponding owner rank
+ *
+ * @param rank The rank under consideration
+ * @param ranks_per_owner Ratio of total ranks to owner ranks
+ * @return The corresponding owner of this ranks
+ * */
+label compute_owner_rank(label rank, label ranks_per_owner);
+
+/* @brief Given a comm_pattern this function executes the communication
+ *
+ * The purpose of this function is to have a uniform function for communication
+ * calls so that we can inspect the communication for debugging purposes
+ *
+ * @param exec_handler The executor handler
+ * @param comm_pattern The communication pattern / send recv counts and offsets
+ * @param send_buffer Pointer to the data buffer to send
+ * @param recv_buffer Pointer to the data buffer to recv
+ */
+void communicate_values(const ExecutorHandler &exec_handler,
+                        const CommCounts &comm_pattern,
+                        const scalar *send_buffer, scalar *recv_buffer);
+
+/* @brief A function that can perform an offseted gather to owner rank
+ *
+ * This function is similar to communicate_values. However, it takes care of
+ * allocating the revc_buffer on the owner rank and can add a prescribed offset
+ * to all elements. The latter is often required when gathering indices to the
+ * owner rank.
+ *
+ * @param exec_handler The executor handler
+ * @param send_buffer Pointer to send data
+ * @param send_size How many elements to send to owner
+ * @param offset Value to add to all elements in the send_buffer
```

Comment:
offset here is not the displacement? but is the value to increment all elements in the send_buffer?

Reply:
yes